File: frontend/lib/models/booking.dart
class Booking {
  final String id;
  final String pnr;
  final String trainId;
  final String trainName;
  final String from;
  final String to;
  final DateTime date;
  final String classType;
  final String status;
  final List<Passenger> passengers;
  final double totalAmount;

  Booking({
    required this.id,
    required this.pnr,
    required this.trainId,
    required this.trainName,
    required this.from,
    required this.to,
    required this.date,
    required this.classType,
    required this.status,
    required this.passengers,
    required this.totalAmount,
  });

  factory Booking.fromJson(Map<String, dynamic> json) {
    return Booking(
      id: json['_id'] ?? json['id'] ?? '',
      pnr: json['pnr'] ?? '',
      trainId: json['trainId'] ?? '',
      trainName: json['trainName'] ?? '',
      from: json['from'] ?? '',
      to: json['to'] ?? '',
      date: DateTime.parse(json['date']),
      classType: json['classType'] ?? '',
      status: json['status'] ?? '',
      passengers: (json['passengers'] as List<dynamic>?)
              ?.map((p) => Passenger.fromJson(p))
              .toList() ??
          [],
      totalAmount: (json['totalAmount'] ?? 0).toDouble(),
    );
  }
}

class Passenger {
  final String name;
  final int age;
  final String gender;

  Passenger({
    required this.name,
    required this.age,
    required this.gender,
  });

  factory Passenger.fromJson(Map<String, dynamic> json) {
    return Passenger(
      name: json['name'] ?? '',
      age: json['age'] ?? 0,
      gender: json['gender'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'age': age,
      'gender': gender,
    };
  }
}

File: backend/src/models/Booking.ts
import mongoose, { Schema, Document } from 'mongoose';

export interface IPassenger {
  name: string;
  age: number;
  gender: 'male' | 'female' | 'other';
  seatPreference?: string;
}

export interface IBooking extends Document {
  pnr: string;
  trainId: string;
  userId: string;
  passengers: IPassenger[];
  classType: 'sleeper' | 'ac3' | 'ac2' | 'ac1';
  date: Date;
  status: 'confirmed' | 'waitlisted' | 'cancelled';
  totalAmount: number;
  createdAt: Date;
  updatedAt: Date;
}

const PassengerSchema = new Schema<IPassenger>({
  name: { type: String, required: true },
  age: { type: Number, required: true },
  gender: { type: String, enum: ['male', 'female', 'other'], required: true },
  seatPreference: { type: String },
});

const BookingSchema = new Schema<IBooking>(
  {
    pnr: { type: String, required: true, unique: true },
    trainId: { type: String, required: true },
    userId: { type: String, required: true },
    passengers: { type: [PassengerSchema], required: true },
    classType: {
      type: String,
      enum: ['sleeper', 'ac3', 'ac2', 'ac1'],
      required: true,
    },
    date: { type: Date, required: true },
    status: {
      type: String,
      enum: ['confirmed', 'waitlisted', 'cancelled'],
      default: 'confirmed',
    },
    totalAmount: { type: Number, default: 0 },
  },
  {
    timestamps: true,
  }
);

export default mongoose.model<IBooking>('Booking', BookingSchema);
